### PR TITLE
initial imex-runner

### PR DESCRIPTION
**Pull request by fschlimb**
Please review these guidelines to help with the review process:
- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, a reproducer, or a reference to an issue with a reproducer?
- [ ] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [x] If this PR is a work in progress, are you filing the PR as a draft?
- [x] Have you organized your commits logically and ensured each can be built by itself?

Adding runner for IMEX. All GPU related features disabled for now.

Trying to get a minimal example working. Materialization and legalization in PTensorToLinalg pass somehow does not work.

func.func @test_arange(%arg0: i64, %arg1: i64, %arg2: i64) -> !ptensor.ptensor<tensor<?xi64>, 1> {
    %0 = "ptensor.arange"(%arg0, %arg1, %arg2) {dist = true} : (i64, i64, i64) -> !ptensor.ptensor<tensor<?xi64>, 1>
    return %0 : !ptensor.ptensor<tensor<?xi64>, 1>
}
@Hardcode84 Could you please help? I sent you more data elsewhere.
@nbpatel

### CODE REVIEW

**Comment by fschlimb** [comment]
Thanks @Hardcode84 for your hints.

The below code still gives an error:
```
Error: only single f32 function result supported
```
Most probably about the following:
```
%3 = llvm.call @arange(%2, %1, %0) : (i64, i64, i64) -> !llvm.struct<(ptr<i64>, ptr<i64>, i64, array<1 x i64>, array<1 x i64>)>
```

@Hardcode84 @nbpatel  Any idea?

```
module {
    func.func @arange(%arg0: i64, %arg1: i64, %arg2: i64) -> !ptensor.ptensor<tensor<?xi64>, 1> {
        %0 = "ptensor.arange"(%arg0, %arg1, %arg2) {dist = true} : (i64, i64, i64) -> !ptensor.ptensor<tensor<?xi64>, 1>
        return %0 : !ptensor.ptensor<tensor<?xi64>, 1>
    }

    func.func @main() {
        %0 = arith.constant 0 : i64
        %1 = arith.constant 10 : i64
        %2 = arith.constant 2 : i64
        %3 = call @arange(%0, %1, %2) : (i64, i64, i64) -> !ptensor.ptensor<tensor<?xi64>, 1>
        %4 = "builtin.unrealized_conversion_cast"(%3) : (!ptensor.ptensor<tensor<?xi64>, 1>) -> tensor<?xi64>
        call @printMemrefF32(%4) : (tensor<?xi64>) -> ()
        return
    }

    func.func private @printMemrefF32(%ptr : tensor<?xi64>)
}
```

**Comment by fschlimb** [comment]
I fixed the issue above by adding the appropriate command-line args to the runner.

**Review by Hardcode84 — APPROVED** [review]
LGTM, but let's rename it to smth like `ptensor-runner`, as it have hardcoded `ptensor` and `dist` passes and is not a general runner.

**Comment by fschlimb** [comment]
> I'm generally ok with that, bit let's rename it to smth like `ptensor-runner`, as it have hardcoded `ptensor` and `dist` passes and is not a general runner.

My intent was bo make this general 
* add other passes when they get pulled into refactor
* add command-line flags like "--device=[gpu|cpu]" or alike to switch between cpu and gpu runners.

@Hardcode84  Is there anything else you see missing for a generic IMEX runner?

File: test/imex-runner/ptensor_arange.mlir [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    func.func @main() {
+        %0 = arith.constant 0 : i64
+        %1 = arith.constant 10 : i64
+        %2 = arith.constant 2 : i64
+        %3 = call @arange(%0, %1, %2) : (i64, i64, i64) -> (!ptensor.ptensor<tensor<?xi64>, 0>)
+
+        %4 = "builtin.unrealized_conversion_cast"(%3) : (!ptensor.ptensor<tensor<?xi64>, 0>) -> tensor<?xi64>
+        %unranked = tensor.cast %4 : tensor<?xi64> to tensor<*xi64>
+        call @printMemrefI64(%unranked) : (tensor<*xi64>) -> ()
+        // CHECK: Unranked Memref base@ = {{0x[-9a-f]*}}
```
[the remark this answers is not in the document]
Comment (fschlimb) [reply test/imex-runner/ptensor_arange.mlir]: fixed.

**Comment by Hardcode84** [comment]
> * add other passes when they get pulled into refactor
> 
> * add command-line flags like "--device=[gpu|cpu]" or alike to switch between cpu and gpu runners.

This doesn't make it 'general' as it still have (mostly) hardcoded pass sequence.
It is ok and convenient to have such tool but it is not generic enough to be named `imex-runner`

If we want a generic `imex-runner` it should not have any hardcoded passes, either taking pass pipeline as command line parameter or expecting already processed IR as input (e.g. plumbed through `imex-opt` first)

Let's just rename this to `ptensor-runner` for now

**Comment by fschlimb** [comment]
> > * add other passes when they get pulled into refactor
> > * add command-line flags like "--device=[gpu|cpu]" or alike to switch between cpu and gpu runners.
> 
> This doesn't make it 'general' as it still have (mostly) hardcoded pass sequence. It is ok and convenient to have such tool but it is not generic enough to be named `imex-runner`
> 
> If we want a generic `imex-runner` it should not have any hardcoded passes, either taking pass pipeline as command line parameter or expecting already processed IR as input (e.g. plumbed through `imex-opt` first)
> 
> Let's just rename this to `ptensor-runner` for now

I would prefer starting with the right approach right away instead of introducing temporaries.

My latest change moves imex-passes to imex-opt and only uses power passes. the test now first runs imex-opt and feeds into imex-runner.

I am not sure I really understand your point of hard-coded passes. We do not really need any runner if all we do is run LLVM code. I thought the whole point of the runner is to provide a pipeline of lower level passes that IMEX is based on to avoid duplication of this lower level stuff. Is this not what it should be? Do you want a runner which accepts LLVM dialect only? That would make our test very verbose.

**Comment by Hardcode84** [comment]
Even without imex passes it is not generic (and configurable) enough. I do not want to enforce any pipeline for such generic name as `imex-runner`, as imex is just a "A staging ground for MLIR dialects and tools. ". If we want to test specific pipeline we need to either make  `imex-runner` entirely configurable (which actually doesn't have much value at the moment), or to have specifically configured (and named) standalone runner (which is perfectly fine even if we have a generic runner).

Lets just do `ptensor-runner` for now.

**Comment by fschlimb** [comment]
How would such a generic runner look like?

**Comment by fschlimb** [comment]
@Hardcode84 @nbpatel Just made imex-runner a python script using imex-opt and mlir-runner. It accepts a string or file defining the pass pipeline. No more hard-coded passes. This is used in the new test.

**Review by Hardcode84 — APPROVED** [review]

**Pull request merged**